h2o-automl/src/main/java/ai/h2o/automl/AutoML.java

```diff
@@ -225,40 +226,95 @@ public AutoML(Key<AutoML> key, Date startTime, AutoMLBuildSpec buildSpec) {
-            (buildSpec.build_control.stopping_criteria.seed() == -1 ? " (random)" : ""));
+    handleEarlyStoppingParameters(buildSpec);
+
+    String sort_metric = buildSpec.input_spec.sort_metric == null ? null : buildSpec.input_spec.sort_metric.toLowerCase();
+    // TODO: does this need to be updated?  I think its okay to pass a null leaderboardFrame
+    leaderboard = Leaderboard.getOrMakeLeaderboard(projectName(), userFeedback, this.leaderboardFrame, sort_metric);
+
+    this.jobs = new ArrayList<>();
+    this.tempFrames = new ArrayList<>();
+  }
+
+  private void performAutoFeatureEngineering() {
+    // Hardcoded default startegy for now
+    TEApplicationStrategy defaultTEApplicationStrategy = new AllCategoricalTEApplicationStrategy(this.trainingFrame);
+    performAutoTargetEncoding(defaultTEApplicationStrategy);
+  }
+
+  //TODO maybe to minimise coupling it is better to introduce AutoMLTEEncodingHelper that will take AutoML instance and return transformed one
+  private void performAutoTargetEncoding(TEApplicationStrategy strategy) {
+    
+    boolean hasCategoricalVecs = false;
+    for(Vec vec : trainingFrame.vecs()) {
+      if(vec.isCategorical()) {
+        hasCategoricalVecs = true;
+        break;
+      }
+    }
+    
+    if(hasCategoricalVecs) {
+
+      //TODO Either perform random grid search over parameters or introduce evolutionary selection algo
+      BlendingParams blendingParams = new BlendingParams(5, 1);
+      String[] columnsToEncode = strategy.getColumnsToEncode();
+      boolean withBlendedAvg = true;
+      boolean imputeNAsWithNewCategory = true;
+      int seed = 1234; // TODO make it a parameter for users to set
+      byte holdoutType = TargetEncoder.DataLeakageHandlingStrategy.KFold;
+
+      TargetEncoder tec = new TargetEncoder(columnsToEncode, blendingParams);
+
+      Frame trainingFrame = getTrainingFrame();
+      String responseColumnName = trainingFrame.name(trainingFrame.find(getResponseColumn()));
+      String foldColumnName = trainingFrame.name(trainingFrame.find(getFoldColumn()));
+
+      Map<String, Frame> encodingMap = tec.prepareEncodingMap(trainingFrame, responseColumnName, foldColumnName, imputeNAsWithNewCategory);
+      switch (holdoutType) {
+        case TargetEncoder.DataLeakageHandlingStrategy.KFold:
+          this.trainingFrame = tec.applyTargetEncoding(trainingFrame, responseColumnName, encodingMap, holdoutType, foldColumnName, withBlendedAvg, imputeNAsWithNewCategory, seed);
+          this.validationFrame = tec.applyTargetEncoding(getValidationFrame(), responseColumnName, encodingMap, TargetEncoder.DataLeakageHandlingStrategy.None, foldColumnName, withBlendedAvg, imputeNAsWithNewCategory, seed);
+          this.leaderboardFrame = tec.applyTargetEncoding(getLeaderboardFrame(), responseColumnName, encodingMap, TargetEncoder.DataLeakageHandlingStrategy.None, foldColumnName, withBlendedAvg, imputeNAsWithNewCategory, seed);
+          break;
```

Comment:
I don't know how other strategies work, but do you think it would make sense to have this done in 2 steps?
The reason is that I see a lot of duplicate params between from `prepareEncodingMap` to all those `applyTargetEncoding`, for example:
```
TETask task = tec.setup(trainingFrame, responseColumnName, foldColumnName, imputeNAsWithNewCategory, withBlendedAvg, seed);
this.trainingFrame = task.apply(trainingFrame, strategy);
this.validationFrame = task.apply(validationFrame, strategy2);
...
```
maybe the fact that non-training frames always have None strategy could be implicit, dunno...
wdyt?

Reply:
I had an attempt to move some common parameters to the TE class. Use can see with an example of BlendingParams class that it is not being passed twice. But the problem is that we want this two-methods type of API for Target Encoding and we don't have a stateful backend/sessions. When we call from the clients the first method and instantiate TargetEncoding class we can keep it and connect to the second/following calls from the client. For AutoML case it could be possible (and would be a right thing to do) bc there is no communication with the client during TE phase.

Reply:
Regarding None and none-training frames. There is no special mechanism for now to separate training and non-training frames. I had a parameter/flag that is close to this but we decided to remove it. Ideally I would have the following API:  TE.fitAndTranform() - for training and transforming of training frame and TE.transform() - for other frames.  But again the first method would have to return frame and not the encoding map..... but we need it to pass with the second method's call. Of course we could return both encoding map and encoded training frame but this would complicate things with serialisation before sending to the client and also it is quite uncommon for users to expect tuples as a return value.